File: sorting/problem164/problem164.py
from typing import List


class Solution:
    def maximumGap(self, nums: List[int]) -> int:
        if len(nums) < 2:
            return 0
        # LSD radix sort by bytes: O(4 * (n + 256)) time, O(n) extra space
        arr = nums[:]
        for shift in range(0, 32, 8):
            buckets: list[list[int]] = [[] for _ in range(256)]
            for x in arr:
                buckets[(x >> shift) & 0xFF].append(x)
            arr = []
            for b in buckets:
                arr.extend(b)
        return max(arr[i + 1] - arr[i] for i in range(len(arr) - 1))

File: sorting/problem164/problem164.cpp
#include <algorithm>
#include <vector>

using namespace std;

class Solution {
public:
    int maximumGap(vector<int>& nums) {
        if (nums.size() < 2) {
            return 0;
        }
        vector<int> arr = nums;
        for (int shift = 0; shift < 32; shift += 8) {
            vector<vector<int>> buckets(256);
            for (int x : arr) {
                buckets[(x >> shift) & 0xFF].push_back(x);
            }
            arr.clear();
            for (auto& b : buckets) {
                arr.insert(arr.end(), b.begin(), b.end());
            }
        }
        int ans = 0;
        for (size_t i = 1; i < arr.size(); ++i) {
            ans = max(ans, arr[i] - arr[i - 1]);
        }
        return ans;
    }
};
